### PR TITLE
w_tilde_*_from jitted by JAX

### PR DESCRIPTION
**Pull request by ickc**
Discussion points:

- [x] is test needed? Before I created this PR, I basically wrote the original output to a file and check that this function is `allclose` to the original output.
- [x] Currently, `jax.jit` is used directly. Do you want me to point to `numba_util.jit` instead?
- [x] I `import jax.numpy as jnp`, as I need the namespace of `np.ndarray` for type hints. Need change here?

### CODE REVIEW

**Comment by ickc** [comment]
There's some preliminary good news:

Using M = 1000, N = 1000 (defined in the code, basically the 2 arbitrary size of the input arrays), comparing the original implementation, the rewritten implementation jitted by numba, and that jitted by jax, there's non-trivial speed up: (benchmark ran on cosma5 login nodes so it is a rough estimate)

Screenshot 2024-12-24 at 7 27 12 PM

Screenshot 2024-12-24 at 7 27 20 PM

File: w_tilde/w_tilde_curvature.py [hunk 1]
```diff
@@ -1,17 +1,22 @@
...
+@jit
 def w_tilde_curvature_interferometer_from(
-    noise_map_real: np.ndarray,
-    uv_wavelengths: np.ndarray,
-    grid_radians_slim: np.ndarray,
-) -> np.ndarray:
+    noise_map_real: np.ndarray[tuple[int], np.float64],
+    uv_wavelengths: np.ndarray[tuple[int, int], np.float64],
+    grid_radians_slim: np.ndarray[tuple[int, int], np.float64],
+) -> np.ndarray[tuple[int, int], np.float64]:
```
Comment (rhayes777) [inline w_tilde/w_tilde_curvature.py]: We're still supporting older versions of Python so could we use type annotations from the typing library?

e.g.
```python
from typing import Tuple

Tuple[int]

**Comment by Jammy2211** [comment]
>is test needed? Before I created this PR, I basically wrote the original output to a file and check that this function is allclose to the original output.

Unit tests are not needed, there are unit tests for these functions in the source code which will be invoked when they get moved there.

> Currently, jax.jit is used directly. Do you want me to point to numba_util.jit instead?

For now, its fine to use `jax.jit`, we will complete the full implementation of the likelihood function outside the source code here and then move things over later.

> I import jax.numpy as jnp, as I need the namespace of np.ndarray for type hints. Need change here?

Again, fine for now, I guess once the source code is JAX only we can just use `jnp.ndarray`?

File: w_tilde/w_tilde_curvature.py [hunk 2]
```diff
@@ -23,48 +28,46 @@ def w_tilde_curvature_interferometer_from(
     `w_tilde_preload_interferometer_from` describes a compressed representation that overcomes this hurdles. It is
     advised `w_tilde` and this method are only used for testing.
 
+    .. math::
+        W̃_{ij} = \sum_{k=1}^N \frac{1}{n_k^2} \cos(2\pi[(g_{i1} - g_{j1})u_{k0} + (g_{i0} - g_{j0})u_{k1}])
+
     Parameters
     ----------
-    noise_map_real
+    noise_map_real : ndarray, shape (N,), dtype=float64
```
Comment (Jammy2211) [inline w_tilde/w_tilde_curvature.py]: Our current convention is to not include typing in the "Parameters" numpy style docstring.

For building `readthedocs`, I use a tool to automatically use the main function typing to perform documentation typing.

In my opinion, including the typing here repeats the typing which is fully visible in the function constructor so prefer to remove it.

File: w_tilde/w_tilde_curvature.py [hunk 3]
```diff
@@ -28,6 +28,9 @@ def w_tilde_curvature_interferometer_from(
     `w_tilde_preload_interferometer_from` describes a compressed representation that overcomes this hurdles. It is
     advised `w_tilde` and this method are only used for testing.
 
+    Note that the current implementation does not take advantage of the fact that w_tilde is symmetric,
```
Comment (Jammy2211) [inline w_tilde/w_tilde_curvature.py:36]: There is a repeating pattern of values in w_tilde which a big question in the future will be if we can exploit within the constraints of JAX.

We can discuss this in the next few weeks, once we have a more end-to-end calculation going.

**Review by Jammy2211 — APPROVED** [review]
Really impressed how quickly and easily you did these conversions, you are clearly a JAX / numpy expert!

File: w_tilde/w_tilde_curvature.py (hunk 1 above)
Comment (ickc) [reply w_tilde/w_tilde_curvature.py]: Actually with `from __future__ import annotations` in the beginning, introduced in Python 3.7, it will "postpone evaluation of annotations" and because of that, it won't cause any runtime errors. I.e. you can use any kind of type annotation futures, even ones that is not supported by the current Python version, and the type checker of choice can still process it.

The caveat is that you need to add `from __future__ import annotations` in the beginning of the `.py` file when you add this function there. This is actually a good practice, as it would also improve performance (without it, the type annotation would be evaluated at runtime which would take a little bit of time.) This however usually is a small effect for scientific computing as the numerical calculations usually takes so much longer.

Comment (ickc) [reply w_tilde/w_tilde_curvature.py]: Resolve in ac1ed27

File: w_tilde/w_tilde_curvature.py (hunk 2 above)
Comment (ickc) [reply w_tilde/w_tilde_curvature.py]: Resolve in 634037.

File: w_tilde/w_tilde_curvature.py (hunk 3 above)
Comment (ickc) [reply w_tilde/w_tilde_curvature.py:36]: Taking advantage of the symmetry would be quite an undertaking and might not be beneficial as there are tradeoffs. Basically in order to not repeat the computation, you need to skip some components and possibly revisit it later, this breaks cache locality and depending on if your computation is intensive enough for this to be worth the extra overhead. With the current function, the only non-trivial compute is the cosine, which is still not very expensive to calculate, so it is possible for it to be slower.

(Breaking the array into tiles is a standard technique to improve cache locality in this kind of operation but then implementing this portably across CPU and GPU, together with the limitation from JAX, can be daunting.)

Also, if we want to take advantage of the symmetry effectively across consistently, we may want to implement an array structure that, say, stores only the upper diagonal of the matrix, to save memory need by about a factor of 2.

Comment (Jammy2211) [reply w_tilde/w_tilde_curvature.py:36]: Yep, I agree, I think the plan for now is to get an end-to-end implementation working using this approach and profiling it to assess if its fast enough for the science we want to do.

There is definitely a limitation that `w_tilde`, which has shape `[image_pixels, image_pixels]`, begins to take up huge amounts of memory because `image_pixels` can exceed values of 80000. Therefore we will probably need to come back to this with a view to reduce memory use, but that may well involve something simpler than fully exploiting the repeating pattern.
